rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,136 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
-                                        *e,
-                                    ),
-                                );
-                            }
-                            RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
-                                error!("Error creating Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentCreationError(
-                                        *e,
-                                    ),
-                                );
-                            }
-                        }
-                    }
-                };
-                for merged_id in merged_ids {
-                    // Skip already taken from log.
-                    if visited_ids.contains(&merged_id) {
-                        continue;
-                    }
-
-                    let user_id = match record_segment_reader
-                        .get_user_id_for_offset_id(merged_id)
-                        .await
-                    {
-                        Ok(user_id) => user_id,
-                        Err(e) => {
-                            println!("Error reading Record Segment: {:?}", e);
-                            return Err(MergeMetadataResultsOperatorError::RecordSegmentReadError);
-                        }
-                    };
-                    ids.push(user_id.to_string());
+        let mut logged_offset_ids: HashSet<u32> = HashSet::new();
+        if input.include_metadata {
+            metadata = vec![None; take_count];
+            documents = vec![None; take_count];
+        }
 
-                    if input.include_metadata {
-                        let record = match record_segment_reader
-                            .get_data_for_offset_id(merged_id)
-                            .await
-                        {
-                            Ok(record) => record,
-                            Err(e) => {
-                                tracing::error!("Error reading Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentReadError,
-                                );
-                            }
-                        };
-                        metadata.push(record.metadata.clone());
-                        match record.document {
-                            Some(document) => documents.push(Some(document.to_string())),
-                            None => documents.push(None),
-                        }
-                    }
+        // Hydrate the data from the materialized logs first
+        for (log, _) in mat_records.iter() {
+            // It's important to account for the records that are
+            // deleted also here so that we can subsequently ignore
+            // them when reading the record segment.
+            logged_offset_ids.insert(log.offset_id);
```

Comment:
this seems incorrect. It does not take into account if the log entry is in Deleted state. We should skip those offset ids. See the original code for more details

Reply:
We need to keep track of what `offset_id`s exists in log regardless whether they are deleted, and we ignore all of them in the compacted data because those are outdated. The deleted `offset_id`s will not be present in `truncated_offset_ids`, and this they will not be included in result. From my understanding previous version of this code does the same thing (`visited_ids` keeps track of these `offset_id`s): https://github.com/chroma-core/chroma/blob/efea481e0d5978af6da33509fc13679d28227b3d/rust/worker/src/execution/operators/merge_metadata_results.rs#L193-L214

Reply:
Update: there might be deleted `offset_id`s present in `truncated_offset_ids`

Reply:
Update: I've made sure `truncated_offset_ids` does not contain any deleted `offset_ids`.